### ISSUE TITLE
core.BackupEntry is reconciled multiple times in a row

### ISSUE THREAD

**Issue by timebertt** [gardener/gardener#3672]
**How to categorize this issue?**

/area robustness quality
/kind bug
/priority normal

**What happened**:

In some gardenlet logs I observed, that a `core.BackupEntry` is always reconciled multiple times in a row.

```
time="2021-03-05T06:49:06Z" level=info msg=Succeeded flow="Shoot cluster reconciliation" operation=reconcile shoot=garden-project/some-shoot task="Deploying backup entry"
time="2021-03-05T06:49:06Z" level=info msg="Backup entry has been successfully reconciled." operation=reconcile shoot=garden-project/some-shoot
time="2021-03-05T06:49:06Z" level=info msg=Succeeded flow="Shoot cluster reconciliation" operation=reconcile shoot=garden-project/some-shoot task="Waiting until the backup entry has been reconciled"
time="2021-03-05T06:49:06Z" level=info msg=Starting backupentry=shoot--it--some-shoot--backup-bucket-uid flow="Backup Entry Reconciliation"
time="2021-03-05T06:49:06Z" level=info msg=Succeeded backupentry=shoot--it--some-shoot--backup-bucket-uid flow="Backup Entry Reconciliation" task="Waiting until the backup bucket is reconciled"
time="2021-03-05T06:49:06Z" level=info msg=Succeeded backupentry=shoot--it--some-shoot--backup-bucket-uid flow="Backup Entry Reconciliation" task="Deploying backup entry secret to seed"
time="2021-03-05T06:49:06Z" level=info msg=Succeeded backupentry=shoot--it--some-shoot--backup-bucket-uid flow="Backup Entry Reconciliation" task="Deploying backup entry extension resource"
time="2021-03-05T06:49:06Z" level=info msg=Succeeded backupentry=shoot--it--some-shoot--backup-bucket-uid flow="Backup Entry Reconciliation" task="Waiting until backup entry is reconciled"
time="2021-03-05T06:49:06Z" level=info msg=Finished backupentry=shoot--it--some-shoot--backup-bucket-uid flow="Backup Entry Reconciliation"
time="2021-03-05T06:49:06Z" level=error msg="Failed to ensure gardener finalizer on backupentry: Operation cannot be fulfilled on backupentries.core.gardener.cloud \"shoot--it--some-shoot--backup-bucket-uid\": the object has been modified; please apply your changes to the latest version and try again" backupentry=shoot--it--some-shoot--backup-bucket-uid
time="2021-03-05T06:49:06Z" level=info msg="Error syncing backupentry garden-project/shoot--it--some-shoot--backup-bucket-uid: Operation cannot be fulfilled on backupentries.core.gardener.cloud \"shoot--it--some-shoot--backup-bucket-uid\": the object has been modified; please apply your changes to the latest version and try again"
time="2021-03-05T06:49:06Z" level=info msg=Starting backupentry=shoot--it--some-shoot--backup-bucket-uid flow="Backup Entry Reconciliation"
time="2021-03-05T06:49:06Z" level=info msg=Succeeded backupentry=shoot--it--some-shoot--backup-bucket-uid flow="Backup Entry Reconciliation" task="Waiting until the backup bucket is reconciled"
time="2021-03-05T06:49:06Z" level=info msg=Succeeded backupentry=shoot--it--some-shoot--backup-bucket-uid flow="Backup Entry Reconciliation" task="Deploying backup entry secret to seed"
time="2021-03-05T06:49:06Z" level=info msg=Succeeded backupentry=shoot--it--some-shoot--backup-bucket-uid flow="Backup Entry Reconciliation" task="Deploying backup entry extension resource"
time="2021-03-05T06:49:12Z" level=info msg=Succeeded backupentry=shoot--it--some-shoot--backup-bucket-uid flow="Backup Entry Reconciliation" task="Waiting until backup entry is reconciled"
time="2021-03-05T06:49:12Z" level=info msg=Finished backupentry=shoot--it--some-shoot--backup-bucket-uid flow="Backup Entry Reconciliation"
```

**What you expected to happen**:

If the `BackupEntry` was reconciled once, it should not be reconciled another time directly afterwards.

**How to reproduce it (as minimally and precisely as possible)**:

1. Create a `core.BackupEntry` (e.g. indirectly by creating a `Shoot` on a `Seed` that has backups enabled)
2. Observe the gardenlet's logs.

**Anything else we need to know?**:

**Environment**:

- Gardener version:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- Others:

**Comment by plkokanov** [comment]
Took a bit of a look into this as part of #3838, seems it's because of this code: [strategy.go:84](https://github.com/gardener/gardener/blob/66d5c9aaa0c63109ba03e164c487d1bcd1474c9b/pkg/registry/core/backupentry/strategy.go#L84) 

The `gardener.cloud/operation=reconcile` is kept on the resource until it is completely reconciled so any updates to it will increase the generation and re-add it to the reconciliation queue.

I could fix it but, separately from #3838 if it is not necessary there.

**Comment by rfranzke** [comment]
Interesting @plkokanov, I'm not sure why this is done differently compared to the operation annotation on the `Shoot`. Probably it should be removed right away so that the controller does not have to do it itself?

**Comment by plkokanov** [comment]
Yes I think removing it right away makes sense. As I see for the shoot, the operation annotation is directly removed after the generation of the resource is increased. So I will try that out.

Edit: there is one more reason - seems like the `BackupEntry.Status.ObservedGeneration` is only modified at the end of a successful operation - it should really be modified at the beginning before any other updates are made to the resource.

**Comment by timebertt** [comment]
/assign @plkokanov 
as you said you want to work on this.

**Comment by timebertt** [comment]
Fixed in #3880
/close
